bmoauth.php

```diff
@@ -0,0 +1,242 @@
+<?php
+
+final class PhabricatorBMOAuthProvider
+  extends PhabricatorAuthProvider {
+
+  const TOKEN_TYPE = 'bmo:auth:request';
+  const TRANSACTION_CODE_LENGTH = 32;
+
+  // Need to add this to avoid error during auth addition activation
+  protected $adapter;
+
+  public function getProviderName() {
+    return pht('BMO Auth Delegation');
+  }
+
+  public function getDescriptionForCreate() {
+    return pht(
+      'Configure a connection to bugzilla.mozilla.org so that users may use '.
+      'BMO credentials to log in to Phabricator.');
+  }
+
+  public function getAdapter() {
+    if (!$this->adapter) {
+      $this->adapter = new PhabricatorBMOAuthAdapter();
+    }
+    return $this->adapter;
+  }
+
+  protected function renderLoginForm(AphrontRequest $request, $mode) {
+    $adapter = $this->getAdapter();
+
+    $attributes = array(
+      'method' => 'GET',
+      'uri' => $adapter->getAuthenticateURI(
+        PhabricatorEnv::getURI($this->getLoginURI()).$this->getAuthCSRFCode($request).'/'
+      ),
+    );
+
+    return $this->renderStandardLoginButton($request, $mode, $attributes);
+  }
+
+  public function processLoginRequest(
+    PhabricatorAuthLoginController $controller) {
+
+    $request = $controller->getRequest();
+    $adapter = $this->getAdapter();
+    $account = null;
+    $response = null;
+
+    // When Bugzilla sends us a "back channel" POST during registration...
+    if ($request->isHTTPPost()) {
+      // Get the `client_api_key` and `client_api_login` from Bugzilla's POST data
+      $post_info = array();
+      try {
+        $post_info = phutil_json_decode(PhabricatorStartup::getRawInput());
+      }
+      catch(Exception $ex) {
+        throw new Exception(pht('Phabricator BMO Authentication failed due to invalid JSON from Bugzilla.'));
+      }
+
+      // Throw exception if either key is not provided by Bugzilla
+      if(!isset($post_info['client_api_key']) || !isset($post_info['client_api_login'])) {
+        throw new Exception(pht('Phabricator BMO Authentication failed due to incomplete JSON from Bugzilla.'));
+      }
+
+      // Generate a transaction code which we'll receive back from Bugzilla
+      // To confirm the API and Client Login information which we saved
+      // In PHP7 this becomes: bin2hex(random_bytes(self::TRANSACTION_CODE_LENGTH))
+      $trans_code = bin2hex(openssl_random_pseudo_bytes(self::TRANSACTION_CODE_LENGTH));
```

Comment:
These can be tricky to get the security correct on, we should definitely have this part reviewed. In MozReview we use python's `uuid.uuid4()`